### ISSUE TITLE
predictor.explain: eli5 and errors

### ISSUE THREAD

**Issue by mattivi** [amaiya/ktrain#368]
I am following the example at [this notebook example](https://nbviewer.jupyter.org/github/amaiya/ktrain/blob/master/tutorials/tutorial-A2-explaining-predictions.ipynb#Explaining-Text-Classification) to explain prediction for a text document with top loss.
I get the errors below, and not able to locate the root cause. Any hint? 
Thanks a lot

```predict``` correctly works:
```python
predictor.predict(example_text)
>> [('not_label_1', 1.0), ('label_1', 2.3528226e-11)]
```
```explain``` has the following error:
```python
predictor.explain(example_text)
>>
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-56-e353340ba859> in <module>
----> 1 predictor.explain(example_text)

~/conda/lib/python3.7/site-packages/ktrain/text/predictor.py in explain(self, doc, truncate_len, all_targets, n_samples)
    130         te = TextExplainer(random_state=42, n_samples=n_samples)
    131         _ = te.fit(doc, self.predict_proba)
--> 132         return te.show_prediction(target_names=self.preproc.get_classes(), targets=prediction)
    133 
    134 

~/conda/lib/python3.7/site-packages/eli5/lime/lime.py in show_prediction(self, **kwargs)
    277         self._fix_target_names(kwargs)
    278         return eli5.show_prediction(self.clf_, self.doc_, vec=self.vec_,
--> 279                                     **kwargs)
    280 
    281     def explain_prediction(self, **kwargs):

~/conda/lib/python3.7/site-packages/eli5/ipython.py in show_prediction(estimator, doc, **kwargs)
    305     """
    306     format_kwargs, explain_kwargs = _split_kwargs(kwargs)
--> 307     expl = explain_prediction(estimator, doc, **explain_kwargs)
    308     if expl.image is not None:
    309         # dispatch to image display implementation

~/conda/lib/python3.7/functools.py in wrapper(*args, **kw)
    822                             '1 positional argument')
    823 
--> 824         return dispatch(args[0].__class__)(*args, **kw)
    825 
    826     funcname = getattr(func, '__name__', 'singledispatch function')

~/conda/lib/python3.7/site-packages/eli5/sklearn/explain_prediction.py in explain_prediction_linear_classifier(clf, doc, vec, top, top_targets, target_names, targets, feature_names, feature_re, feature_filter, vectorized)
    190     classes = getattr(clf, "classes_", ["-1", "1"])  # OneClassSVM support
    191     display_names = get_target_display_names(classes, target_names,
--> 192                                              targets, top_targets, score)
    193 
    194     if is_multiclass_classifier(clf):

~/conda/lib/python3.7/site-packages/eli5/utils.py in get_target_display_names(original_names, target_names, targets, top_targets, score)
    173         raise ValueError('Pass either "targets" or "top_targets", not both')
    174 
--> 175     target_indices = _get_value_indices(original_names, display_names, targets)
    176     names = [display_names[i] for i in target_indices]
    177     return list(zip(target_indices, names))

~/conda/lib/python3.7/site-packages/eli5/utils.py in _get_value_indices(names1, names2, lookups)
    230     positions = {name: idx for idx, name in enumerate(names2)}
    231     positions.update({name: idx for idx, name in enumerate(names1)})
--> 232     return [positions[name] for name in lookups]
    233 
    234 

~/conda/lib/python3.7/site-packages/eli5/utils.py in <listcomp>(.0)
    230     positions = {name: idx for idx, name in enumerate(names2)}
    231     positions.update({name: idx for idx, name in enumerate(names1)})
--> 232     return [positions[name] for name in lookups]
    233 
    234 

TypeError: unhashable type: 'list'
```

**Comment by amaiya** [comment]
Based on your output from `predictor.predict`, it looks like your model has been setup for multilabel classification (i.e., labels are not mutually-exclusive).    The libraries used by the `explain` method do not support multilabel classification.  However, the error is normally something like `Probabilities do not sum to 1`:

```python
# multilabel model
p.predict('My computer monitor is blurry.')

# output
# [('alt.atheism', 0.015460613),
#   ('comp.graphics', 0.97166234),
 #  ('sci.med', 0.029763268),
 #  ('soc.religion.christian', 0.015686162)]
```
```python
p.explain('My computer monitor is blurry.')
# error output
# ValueError: probabilities do not sum to 1
```

Is your dataset a multilabel problem?  What does your dataset look like?

**Comment by mattivi** [comment]
Thanks for very quick reply! It's not a multilabel classification problem, text is in column "text", while labels are numeric type ( 0 and 1) and in column "label_4_metrics". Code here:
```python
(x_train, y_train), (x_valid, y_valid), preproc = text.texts_from_df(train_df=X_train, 
                                                                   text_column='text',
                                                                   label_columns='label_4_metrics',
                                                                   val_df=X_valid,
                                                                   maxlen=MAX_LEN, 
                                                                   max_features=MAX_FEATURES,
                                                                   preprocess_mode='standard',
                                                                   ngram_range=NGRAM_RANGE)
```
that provides
```
['not_label_4_metrics', 'label_4_metrics']
   not_label_4_metrics  label_4_metrics
0                  1.0              0.0
1                  1.0              0.0
2                  1.0              0.0
3                  1.0              0.0
4                  0.0              1.0
['not_label_4_metrics', 'label_4_metrics']
   not_label_4_metrics  label_4_metrics
0                  1.0              0.0
1                  1.0              0.0
2                  1.0              0.0
3                  1.0              0.0
4                  1.0              0.0
language: de
Word Counts: 23281
Nrows: 15080
15080 train sequences
etc etc etc
```

It works with 
```predictor.explain(example_text, all_targets=True)``` but error with all_targets=False

**Comment by amaiya** [comment]
Thanks.  What kind of model are you using? Is it the same one in the notebook example or a different one?

**Comment by mattivi** [comment]
It's a logistic regression:
```python
MAX_FEATURES = 32000
MAX_LEN = 12000
NGRAM_RANGE = 1
BATCH_SIZE = 16
CLASSIFIER = 'logreg'
LR =5e-5

(x_train, y_train), (x_valid, y_valid), preproc = text.texts_from_df(train_df=X_train, 
                                                                   text_column='text',
                                                                   label_columns='label_4_metrics',
                                                                   val_df=X_valid,
                                                                   maxlen=MAX_LEN, 
                                                                   max_features=MAX_FEATURES,
                                                                   preprocess_mode='standard',
                                                                   ngram_range=NGRAM_RANGE)

tst, y_tst = preproc.preprocess_test(X_test["text"].values, y_test=X_test["label_4_metrics"])

model = text.text_classifier(CLASSIFIER, (x_train, y_train) , preproc=preproc)

metric_auc = tf.keras.metrics.AUC()
model.compile(loss="binary_crossentropy", optimizer='adam', metrics=[metric_auc])

learner = ktrain.get_learner(model, 
                             train_data=(x_train, y_train), 
                             val_data=(x_valid, y_valid), 
                             batch_size=BATCH_SIZE)

es_roc = EarlyStopping(monitor='val_auc', mode='max', verbose=1, patience=5)
es_loss = EarlyStopping(monitor='val_loss', mode='min', verbose=1, patience=5)
history = learner.fit(LR, 100, callbacks=[es_roc, es_loss])

predictor = ktrain.get_predictor(learner.model, preproc=preproc)

predictor.explain(example_text, all_targets=False)
```
And 
```python
predictor.get_classes()
>> ['not_label_4_metrics', 'label_4_metrics']
```

**Comment by amaiya** [comment]
Thanks for posting the full code.  I noticed one issue in your code that may be causing problems.

When you run `preprocess_train`, your single column of labels (**label_4_metrics**) is 1-hot-encoded and transformed into an array of shape `(N,2)` (**not_label_4_metrics** and **label_4_metrics**).  This means you need to use `categorical_crossentropy` as the loss function instead of `binary_crossentropy`.  Since you've configured your model to use `binary_crossentropy`, it is being treated by Keras as a multilabel model under the hood.  This may be causing problems for **eli5** which is used by `explain`.

I would recommend changing the loss function to `categorical_crossentropy` and see what happens with `explain`.

**Comment by mattivi** [comment]
Correct, now it works changing into ```categorical_crossentropy```. Thanks!
